subsys/net/lib/nrf_cloud/src/nrf_cloud_client_id.c

```diff
@@ -0,0 +1,115 @@
+#include <modem/modem_jwt.h>
+#endif
+#include <modem/at_cmd.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <logging/log.h>
+#include <sys/util.h>
+#include "nrf_cloud_transport.h"
+
+LOG_MODULE_REGISTER(nrf_cloud_client_id, CONFIG_NRF_CLOUD_LOG_LEVEL);
+
+#if defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_COMPILE_TIME)
+BUILD_ASSERT((sizeof(CONFIG_NRF_CLOUD_CLIENT_ID) - 1) <= NRF_CLOUD_CLIENT_ID_MAX_LEN,
+	"CONFIG_NRF_CLOUD_CLIENT_ID must not exceed NRF_CLOUD_CLIENT_ID_MAX_LEN");
+#endif
+
+#if defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI)
+#define CGSN_RESPONSE_LENGTH 19
+#define NRF_IMEI_LEN 15
+#define IMEI_CLIENT_ID_LEN (sizeof(CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX) \
+			    - 1 + NRF_IMEI_LEN)
+BUILD_ASSERT(IMEI_CLIENT_ID_LEN <= NRF_CLOUD_CLIENT_ID_MAX_LEN,
+	"NRF_CLOUD_CLIENT_ID_PREFIX plus IMEI must not exceed NRF_CLOUD_CLIENT_ID_MAX_LEN");
+#endif
+
+int nrf_cloud_client_id_get(char *id_buf, size_t id_len)
+{
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+	return nct_client_id_get(id_buf, id_len);
+#else
+	return nrf_cloud_cfgd_client_id_get(id_buf, id_len);
+#endif
+}
+
+size_t nrf_cloud_cfgd_client_id_length_get(void)
+{
+#if defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI)
+	return IMEI_CLIENT_ID_LEN;
+#elif defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_INTERNAL_UUID)
+	return NRF_DEVICE_UUID_STR_LEN;
+#elif defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_COMPILE_TIME)
+	return (sizeof(CONFIG_NRF_CLOUD_CLIENT_ID) - 1);
+#else
+	return 0;
+#endif
+}
+
+int nrf_cloud_cfgd_client_id_get(char * const buf, const size_t buf_sz)
+{
+	if (!buf || !buf_sz) {
+		return -EINVAL;
+	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME)) {
+		LOG_WRN("Configured for runtime client ID");
+		return -ENODEV;
+	}
+
+	int err;
+	int print_ret;
+
+#if (defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_INTERNAL_UUID) || \
+	defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI))
+	/* UUID/IMEI are obtained via AT command */
+	if (!IS_ENABLED(CONFIG_AT_CMD_SYS_INIT)) {
+		err = at_cmd_init();
+		if (err) {
+			LOG_ERR("at_cmd failed to initialize, error: %d", err);
+			return err;
+		}
+	}
+#endif
+
+#if defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI)
+	char imei_buf[CGSN_RESPONSE_LENGTH + 1];
+
+	err = at_cmd_write("AT+CGSN", imei_buf, sizeof(imei_buf), NULL);
+	if (err) {
+		LOG_ERR("Failed to obtain IMEI, error: %d", err);
+		return err;
+	}
+
+	imei_buf[NRF_IMEI_LEN] = 0;
+
+	print_ret = snprintf(buf, buf_sz, "%s%.*s",
+			     CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX,
+			     NRF_IMEI_LEN, imei_buf);
+
+#elif defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_INTERNAL_UUID)
+	struct nrf_device_uuid dev_id;
+
+	err = modem_jwt_get_uuids(&dev_id, NULL);
+	if (err) {
+		LOG_ERR("Failed to get device UUID: %d", err);
+		return err;
+	}
+
+	print_ret = snprintf(buf, buf_sz, "%s", dev_id.str);
+
+#elif defined(CONFIG_NRF_CLOUD_CLIENT_ID_SRC_COMPILE_TIME)
+	print_ret = snprintf(buf, buf_sz, "%s", CONFIG_NRF_CLOUD_CLIENT_ID);
+#endif
+
+	if (print_ret <= 0 || print_ret >= buf_sz) {
+		return -EIO;
```

Comment:
i suppose this would be more accurate:
```
if (print_ret <= 0) {
		return -EIO;
} else if (print_ret >= buf_sz) {
		return -EMSGSIZE;
}
```
or even -ENOBUFS? 

looking at other places in the nrf_cloud that do a similar check, -ETXTBSY is used in nrf_cloud_rest and -EIO is used in allocate_and_format_topic :sweat_smile:

preference?

Reply:
-EMSGSIZE was in the older `nct_client_id_get()`...

E2BIG seems wrong to me if it really means "argument list too long":
#define	E2BIG 7		/* Arg list too long */

ETXTBSY likewise doesn't seem right, if the GNU header is "correct":
#define	ETXTBSY 26	/* Text file busy */